### PR TITLE
Add class weights support for handling class imbalance in training

### LINKED ISSUE #21082: Support class-wise loss weighting for handling dataset imbalance
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

User requests to add configurable per-class weights

### Use case

Compensate for class imbalance in dataset

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by ahmet-f-gumustas**
Adds optional class weights support for imbalanced datasets via a new `cls_pw` hyperparameter.

When enabled, classes with fewer samples get higher weights in the BCE classification loss, helping the model pay more attention to underrepresented classes. Weights are computed as inverse frequency from dataset labels.

- `cls_pw=0.0` → disabled (default, no behavior change)
- `cls_pw=1.0` → full inverse frequency weighting
- `cls_pw=0.5` → partial weighting (square root of inverse frequency)

```bash
yolo train data=my_data.yaml model=yolo11n.pt cls_pw=1.0
```

**Files changed:**
- `ultralytics/utils/ops.py` — `labels_to_class_weights()` function
- `ultralytics/cfg/default.yaml` — `cls_pw: 0.0` parameter
- `ultralytics/cfg/__init__.py` — register `cls_pw` in `CFG_FLOAT_KEYS`
- `ultralytics/engine/trainer.py` — `set_class_weights()` hook
- `ultralytics/models/yolo/detect/train.py` — `set_class_weights()` implementation
- `ultralytics/utils/loss.py` — apply class weights in `v8DetectionLoss`

Tested on COCO dataset with `cls_pw=0.0` and `cls_pw=1.0`, both run successfully.

> **Note:** This PR goes slightly over the 50-line guideline since it introduces an optional end-to-end feature (config + trainer + loss). The changes are scoped and isolated to class-weight support. If preferred, I'm happy to split this into smaller PRs or adjust the implementation.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds optional class-frequency-based weighting to the YOLO detection classification loss to better handle imbalanced datasets ⚖️

### 📊 Key Changes
- Introduces new training hyperparameter `cls_pw` (default `0.0`) in `default.yaml` and registers it in `cfg/__init__.py`.
- Adds a `Trainer.set_class_weights()` hook and calls it after the train dataloader is created.
- Implements detection-specific class-weight computation in `ultralytics/models/yolo/detect/train.py` using inverse class frequency, with optional power scaling and renormalization.
- Applies `model.class_weights` inside `ultralytics/utils/loss.py` by weighting the per-class BCE classification loss.
- Adds `labels_to_class_weights()` utility to `ultralytics/utils/ops.py` for computing normalized inverse-frequency weights.

### 🎯 Purpose & Impact
- Improves training behavior on imbalanced detection datasets by increasing emphasis on underrepresented classes 📈
- Keeps default behavior unchanged (`cls_pw: 0.0` disables weighting), minimizing risk for existing users ✅
- Provides a reusable Trainer hook for future task-specific implementations (e.g., segment/pose) 🔧

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ahmet-f-gumustas, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness—an engineer will take a look soon 🧩

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — DISMISSED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

PR looks clean. The changes are coherent, defaults preserve existing behavior, and the new class-weighting path is well-contained.

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

PR looks clean. Class-weights flow is wired end-to-end without obvious bugs, and the loss application is consistent with existing BCE usage.

**Comment by ahmet-f-gumustas** [comment]
@glenn-jocher Friendly ping :)

**Comment by glenn-jocher** [comment]
Thanks for the ping — we’ll review when bandwidth allows, but I can’t give a merge ETA; at a high level we’d prefer this to stay minimal and avoid adding new `default.yaml` args where possible, since class-weighted loss is already supported through the [custom trainer guide](https://docs.ultralytics.com/guides/custom-trainer/), so if you’d like to keep iterating please align with that pattern and include focused tests/docs.

**Review by Laughing-q — APPROVED** [review]

**Review by Laughing-q — APPROVED** [review]

**Comment by Laughing-q** [comment]
trained yolo26s on coco with epochs=300 from scratch, posting some mAP for rare classes
### baseline

```text
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):
              keyboard        106        153      0.689      0.641      0.709      0.552
            cell phone        214        262      0.643      0.488      0.545       0.38
             microwave         54         55      0.705      0.695      0.746      0.624
                  oven        115        143      0.614      0.545      0.608      0.421
                  sink        187        225      0.652      0.532      0.587      0.407
          refrigerator        101        126      0.827      0.698      0.782      0.665
                 clock        204        267      0.749      0.685      0.709      0.511
                  vase        137        274      0.626      0.522       0.55      0.395
              scissors         28         36      0.594      0.389      0.413      0.361
            teddy bear         94        190       0.72      0.642      0.704      0.544
            hair drier          9         11          1          0      0.114     0.0609
```

### `cls_pw=0.25`

```text
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):
              keyboard        106        153      0.684      0.622      0.707       0.54
            cell phone        214        262      0.661      0.485      0.546      0.373
             microwave         54         55      0.717      0.745      0.813      0.655
                  oven        115        143      0.629      0.524      0.582      0.422
                  sink        187        225      0.704      0.604      0.625       0.44
          refrigerator        101        126      0.802      0.706        0.8      0.686
                 clock        204        267      0.762      0.683      0.725      0.521
                  vase        137        274      0.644      0.555      0.567      0.405
              scissors         28         36       0.79      0.389      0.473      0.391
            teddy bear         94        190      0.732      0.647      0.712      0.548
            hair drier          9         11      0.509     0.0909      0.191      0.121
```
Summary:
- Got improvement on many rare classes, especially for hair drier mAP bumped from `0.06` to `0.121`
- Also got 0.2 mAP improvement overall. baseline: 46.4 vs `cls_pw=0.25`: 46.6

**Comment by Laughing-q** [comment]
@ahmet-f-gumustas Awesome work on this! Thank you for your contribution!

**Comment by ahmet-f-gumustas** [comment]
@Laughing-q Thank You for review
